contributors/design-proposals/sig-cli/kubectl_plan.md

```diff
@@ -0,0 +1,164 @@
+# Kubectl Plan
+
+## TL;DR
+
+Introduce a new `plan` command similar to `kubectl create --dry-run -o yaml`
+
+- Uses server to convert *flags* to a canonical object representation
+  - Client convert flags -> plan request
+  - Server convert plan request -> object
+  - Server returns object without storing
+- `plan` subcommands and help discovered through discovery service (location)
+  and openapi (model)
+
+## Abstract
+
+Kubectl supports a number of imperative commands for building Kubernetes resources
+on the commandline without interacting with config directly.  While users have
+embraced the simplicity and familiarity of using these imperative creation commands,
+the commands hide important details about how the system works, and encourage anti-patterns
+of interacting directly with objects in the system instead of through local configuration.
+
+Additionally because the imperative creation commands
+are completely compiled into the client, they cannot support extension APIs or
+behave well in client / server version skewed environments.
+
+Lastly, maintaining creation subcommands requires sustaining engineering efforts
+by SIG cli to keep up with changes made on the server side.  Server side
+changes frequently break the cli creation commands, requiring the SIG cli
+maintainers to track down the cause and sometimes implement a fix.
+
+This document proposes introducing an optional `plan` subresource which converts the
+commandline arguments into an object on the server side through a subresource and
+simplified representation of the object.  The object is not stored as part of the request,
+instead the client must do so through a separate request.
+
+## Goals
+
+- Provide simple cli commands that reflect how the system is designed
+- Enable API extensions to expose cli commands without downloading plugins 
+- Allow implementers of resource APIs to define cli experience without
+  creating PRs in kubectl
+- Reduce sustaining engineering burden on SIG to maintain the correctness
+  of the commands
+
+## Plan CLI
+
+Kubectl will expose the new `plan` and `plan-list` top-level commands.
+
+When a user runs `kubectl plan -h` it will print a help message
+describing the command, and prompt the user to run `plan-list` to view
+the available subcommands.
+
+**Note:** Alternatively, we could just make the help message itself dynamic instead
+of introducing a `plan-list` command.  We should make this consistent with whatever
+we end up doing for `kubectl get`.
+
+`plan` and `plan-list` will use the discovery API to identify all resources with a `plan`
```

Comment:
Re "list" to get a set of versions, kubernetes/kubernetes#53387 captures some of my thoughts on how we need to improve listing supported resources.

I'm inclined to do it via get and describe, special casing if necessary, for maximum user comprehension.

```
$ kubectl get apigroups
$ kubectl get apigroups extensions
$ kubectl get apiversions extensions/v1beta1 # potentially special case this so that you can look up apiversions like a real user
$ kubectl get apiresources
$ kubectl get apiresources statefulsets.v1.apps
$ kubectl describe apiresources/statefulsets.v1.apps
$ kubectl explain --api-version=apps/v1 statefulsets
```

We can reserve those names (we kind of already have) and if a third party declares their own they can still use subsetting.

Reply:
Like the direction.  Started to read through the issue.

nit: I don't love the arguments and flags differ between the various commands.  The lack of consistency makes it hard to remember the format for each command.  e.g. `--api-version apps/v1 statefulsets` (explain) vs apps/v1` (apiversions) vs statefulsets.v1.apps

Happy to talk through more once the concept is agreed upon.

Reply:
How will the user know what is supported for each api?  e.g. is scale supported for `stateful set`?  How would the user know what supports `plan` (`design`)